Improve build docs and process for Windows

peterjc123 opened the pull request:
Fixes #21026.
1. Improve build docs for Windows
2. Change `BUILD_SHARED_LIBS=ON` for Caffe2 local builds
3. Change to out-source builds for LibTorch and Caffe2 (transferred to #21452)

peterjc123 commented:
@pytorchbot rebase this please

ezyang commented:
Waiting for Windows CI to finish

ezyang commented:
Maybe we should just give Windows install its own section in the README? I'm happy to rearrange things if you think so.

peterjc123 commented:
@pytorchbot retest this please

peterjc123 commented:
This test is failing on Windows but I don't think it's caused by this PR.
```
15:36:10 ======================================================================
15:36:10 FAIL: test_Conv2d_groups_nobias_v2 (__main__.TestNN)
15:36:10 ----------------------------------------------------------------------
15:36:10 Traceback (most recent call last):
15:36:10   File "test_nn.py", line 4804, in test_Conv2d_groups_nobias_v2
15:36:10     self.assertEqual(output, torch.cat([output1, output2], 1))
15:36:10   File "C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test1\test\common_utils.py", line 503, in assertEqual
15:36:10     assertTensorsEqual(x, y)
15:36:10   File "C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test1\test\common_utils.py", line 495, in assertTensorsEqual
15:36:10     self.assertLessEqual(max_err, prec, message)
15:36:10 AssertionError: tensor(1.5259e-05, device='cuda:0', dtype=torch.float16,
15:36:10        grad_fn=<MaxBackward1>) not less than or equal to 1e-05 : 
15:36:10 
15:36:10 ----------------------------------------------------------------------
```

ezyang commented:
Yes, Windows was broken this morning. Should be fixed now.

@pytorchbot rebase this please

Excerpts from peterjc123's message of 2019-05-31 09:15:36 -0700:
> This test is failing on Windows but I don't think it's caused by this PR.
> ```
> 15:36:10 ======================================================================
> 15:36:10 FAIL: test_Conv2d_groups_nobias_v2 (__main__.TestNN)
> 15:36:10 ----------------------------------------------------------------------
> 15:36:10 Traceback (most recent call last):
> 15:36:10   File "test_nn.py", line 4804, in test_Conv2d_groups_nobias_v2
> 15:36:10     self.assertEqual(output, torch.cat([output1, output2], 1))
> 15:36:10   File "C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test1\test\common_utils.py", line 503, in assertEqual
> 15:36:10     assertTensorsEqual(x, y)
> 15:36:10   File "C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test1\test\common_utils.py", line 495, in assertTensorsEqual
> 15:36:10     self.assertLessEqual(max_err, prec, message)
> 15:36:10 AssertionError: tensor(1.5259e-05, device='cuda:0', dtype=torch.float16,
> 15:36:10        grad_fn=<MaxBackward1>) not less than or equal to 1e-05 : 
> 15:36:10 
> 15:36:10 ----------------------------------------------------------------------
> ```
>

peterjc123 commented:
@pytorchbot rebase this please

ezyang commented on tools/build_libtorch.py:
```diff
@@ -19,4 +19,4 @@
 
     build_caffe2(version=None, cmake_python_library=None, build_python=False,
-                 rerun_cmake=True, cmake_only=False, build_dir='.')
+                 rerun_cmake=True, cmake_only=False, build_dir='build')
```
This is probably causing the Caffe2 tests to fail; we'll need to fix the paths.

ezyang reviewed (changes requested):
Build failures seem legit

peterjc123 commented:
@ezyang Could you please tell me which part of the cmake script is related to the installation of the `test_jit` binary when doing out-source builds?

mingzhe09088 committed:
change output of ai_pep_format (#21440)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21440

This diff modifies the output format when ai_pep_format is enabled.

Reviewed By: hl475

Differential Revision: D15681042

fbshipit-source-id: df5f2dbb38d1bd866ca7f74ef4e63459d480be6e

Improve build docs for Windows

Fix comments in batch files

More fixes

More fixes

Remove useless piece

Force exit

Revert changes

Remove unrelated changes

peterjc123 commented:
@ezyang I moved that part to #21452, so it won't block CI build anymore.

ezyang commented:
> Could you please tell me which part of the cmake script is related to the installation of the test_jit binary when doing out-source builds?

I'm not sure. Do you want me to look at this more closely? It's probably one of the scripts in `.jenkins` (or possibly `.circleci`)

ezyang commented:
Merge conflict?!

peterjc123 commented:
@ezyang Resolved.
